helix-core/src/test/java/org/apache/helix/integration/rebalancer/TestMixedModeAutoRebalance.java

```diff
@@ -111,112 +109,125 @@ public void beforeClass() throws Exception {
+      Map<String, List<String>> userDefinedPreferenceLists = idealState.getPreferenceLists();
+
+      List<String> newNodes = new ArrayList<>();
+      for (int i = NUM_NODE; i < NUM_NODE + _replica; i++) {
+        String instance = PARTICIPANT_PREFIX + "_" + (START_PORT + i);
+        _gSetupTool.addInstanceToCluster(CLUSTER_NAME, instance);
+
+        // start dummy participants
+        MockParticipantManager participant =
+            new TestMockParticipantManager(ZK_ADDR, CLUSTER_NAME, instance);
+        participant.syncStart();
+        _participants.add(participant);
+        newNodes.add(instance);
+      }
 
-    ResourceConfig resourceConfig =
-        new ResourceConfig.Builder(DB_NAME).setPreferenceLists(userDefinedPreferenceLists).build();
-    _configAccessor.setResourceConfig(CLUSTER_NAME, DB_NAME, resourceConfig);
+      List<String> userDefinedPartitions = new ArrayList<>();
+      for (String partition : userDefinedPreferenceLists.keySet()) {
+        userDefinedPreferenceLists.put(partition, newNodes);
+        userDefinedPartitions.add(partition);
+      }
 
-    TestHelper.verify(() -> {
-      ExternalView ev =
-          _gSetupTool.getClusterManagementTool().getResourceExternalView(CLUSTER_NAME, DB_NAME);
-      if (ev != null) {
-        for (String partition : ev.getPartitionSet()) {
-          Map<String, String> stateMap = ev.getStateMap(partition);
-          if (stateMap.values().contains("ERROR")) {
-            return true;
+      ResourceConfig resourceConfig =
+          new ResourceConfig.Builder(dbName).setPreferenceLists(userDefinedPreferenceLists).build();
+      _configAccessor.setResourceConfig(CLUSTER_NAME, dbName, resourceConfig);
+
+      TestHelper.verify(() -> {
+        ExternalView ev =
+            _gSetupTool.getClusterManagementTool().getResourceExternalView(CLUSTER_NAME, dbName);
+        if (ev != null) {
+          for (String partition : ev.getPartitionSet()) {
+            Map<String, String> stateMap = ev.getStateMap(partition);
+            if (stateMap.values().contains("ERROR")) {
+              return true;
+            }
           }
         }
-      }
-      return false;
-    }, 2000);
-
-    ExternalView ev =
-        _gSetupTool.getClusterManagementTool().getResourceExternalView(CLUSTER_NAME, DB_NAME);
-    IdealState is = _gSetupTool.getClusterManagementTool().getResourceIdealState(CLUSTER_NAME,
-        DB_NAME);
-    validateMinActiveAndTopStateReplica(is, ev, _replica, NUM_NODE);
+        return false;
+      }, 2000);
+
+      ExternalView ev =
+          _gSetupTool.getClusterManagementTool().getResourceExternalView(CLUSTER_NAME, dbName);
+      IdealState is =
+          _gSetupTool.getClusterManagementTool().getResourceIdealState(CLUSTER_NAME, dbName);
+      validateMinActiveAndTopStateReplica(is, ev, _replica, NUM_NODE);
+    } finally {
+      _gSetupTool.getClusterManagementTool().dropResource(CLUSTER_NAME, dbName);
+      _clusterVerifier.verify(5000);
```

Comment:
1. Should we make 5000 a constant?
2. Does verify() really check that the cluster is gone? Shouldn't we explicitly check for the ZNodes and see if they return null when read or false on an exist() call?

Reply:
1. I agree, but I didn't find a good place to put it. If in this file, we only have 2 references. If move it somewhere else, then different test case might have a different requirement. So I'd prefer to keep the test case easier for now.
2. It just needs to ensure the resource is gone.

Reply:
Okay. As for 2, I actually misread and thought you were doing a dropCluster(). verify makes sense in this case. Since this is a test, I'm fine with keeping 5000 for now, but it would be better to create a timeout constant either inside the verifier or in this class since I do see a lot of magic numbers here.